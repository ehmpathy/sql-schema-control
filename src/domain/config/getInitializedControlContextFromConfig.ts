/*
  1. get the config (validated and hydrated)
  2. initialize the environment based on the config
    - start the db connection
    - initialize the control environment
  3. return the language+dialect, connection, and definitions
*/
