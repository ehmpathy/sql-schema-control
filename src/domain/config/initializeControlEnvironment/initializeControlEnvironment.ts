/*
  complete any actions required to initialize control environment
    1. connect to db
    2. provision change log table
*/
