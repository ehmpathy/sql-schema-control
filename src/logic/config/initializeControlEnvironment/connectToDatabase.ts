import mysql from 'mysql2/promise';
import { DatabaseConnection, DatabaseLanguage, ControlConfig, ConnectionConfig } from '../../../types';

// create the connection to database
const connectionAdapters = {
  [DatabaseLanguage.MYSQL]: async ({ connectionConfig }: { connectionConfig: ConnectionConfig }): Promise<DatabaseConnection> => {
    const connection = await mysql.createConnection({
      host: connectionConfig.host,
      port: connectionConfig.port,
      database: connectionConfig.schema,
      user: connectionConfig.username,
      password: connectionConfig.password,
    });
    return {
      query: async ({ sql, values }: { sql: string, values?: any[] }) => {
        return connection.query(sql, values);
      },
      end: async () => connection.end(),
    };
  },
};

export const connectToDatabase = async ({ config }: { config: ControlConfig }): Promise<DatabaseConnection> => {
  // 1. get the connection adapter for the method
  const getDbConnection = connectionAdapters[config.language];

  // 2. attempt to connect
  const connection = await getDbConnection({ connectionConfig: config.connection });

  // 3. run a test query to check connection
  await connection.query({ sql: 'SHOW TABLES' }); // TODO: make this not mysql specific

  // 4. return the connection
  return connection;
};
