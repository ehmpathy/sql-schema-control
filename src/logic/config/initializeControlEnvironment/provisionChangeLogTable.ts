import { DatabaseConnection } from '../../../types';
import { readFileAsync } from '../_utils/readFileAsync';

export const provisionChangeLogTable = async ({ connection }: { connection: DatabaseConnection }) => {
  // create the table if not exists (sql consideres its existance for us)
  const createTableSql = await readFileAsync({
    filePath: `${__dirname}/../../../../schema/tables/schema_control_change_log.mysql.sql`,
  }); // TODO: make independent of language (mysql currently)
  await connection.query({ sql: createTableSql });
};
