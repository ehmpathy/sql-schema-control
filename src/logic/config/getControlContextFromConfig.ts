import { ControlContext, ResourceDefinition } from '../../types';
import { getUncontrolledResources } from '../schema/resourceDefinition/getUncontrolledResources';
import { getConfig } from './getConfig';
import { initializeControlEnvironment } from './initializeControlEnvironment';

/*
  1. get the config (validated and hydrated)
  2. initialize the environment based on the config
    - start the db connection
    - initialize the control environment
  3. return the control context
*/
export const getControlContextFromConfig = async ({
  configPath,
  strict = false,
}: {
  configPath: string;
  strict?: boolean;
}) => {
  // 1. get the config
  const config = await getConfig({ configPath });

  // 2. initialize the environment based on the config
  const { connection } = await initializeControlEnvironment({ config });

  // 3. define the definitions such that we can append to them if we want (e.g., if want to add uncontrolled)
  const definitions = [...config.definitions];

  // 4. get uncontrolled resources, if strict
  if (config.strict || strict) {
    const controlledResources = definitions.filter(
      (def) => def.constructor === ResourceDefinition,
    ) as ResourceDefinition[];
    const uncontrolledResources = await getUncontrolledResources({ connection, controlledResources });
    definitions.push(...uncontrolledResources); // append the uncontrolled resources to the definitions
  }

  // 5. return the control context
  return new ControlContext({
    language: config.language,
    dialect: config.dialect,
    connection, // note: the connection is a live dbConnection
    definitions, // note: uncontrolled resources will already have statuses
  });
};
