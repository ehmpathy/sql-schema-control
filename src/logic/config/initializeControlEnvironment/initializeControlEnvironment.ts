import { ControlConfig } from '../../../domain';
import { connectToDatabase } from './connectToDatabase';
import { provisionChangeLogTable } from './provisionChangeLogTable';

/*
  complete any actions required to initialize control environment
    1. connect to db
    2. provision change log table
*/
export const initializeControlEnvironment = async ({
  config,
}: {
  config: ControlConfig;
}) => {
  // 1. connect to db
  const connection = await connectToDatabase({ config });

  // 2. provision the change log table
  await provisionChangeLogTable({ connection });

  // 3. return the connection
  return { connection };
};
