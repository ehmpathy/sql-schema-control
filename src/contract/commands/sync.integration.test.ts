import { stdout } from 'stdout-stderr';

import { initializeControlEnvironment } from '../../logic/config/initializeControlEnvironment';
import {
  DatabaseConnection,
  DatabaseLanguage,
  ControlConfig,
} from '../../types';
import { promiseConfig as promiseConfigMysql } from '../__test_assets__/mysql/connection.config';
import { promiseConfig as promiseConfigPostgres } from '../__test_assets__/postgres/connection.config';
import Sync from './sync';

describe('sync', () => {
  describe('mysql', () => {
    let connection: DatabaseConnection;
    beforeAll(async () => {
      const config = new ControlConfig({
        language: DatabaseLanguage.MYSQL,
        dialect: '5.7',
        connection: await promiseConfigMysql(),
        definitions: [],
        strict: false,
      });
      ({ connection } = await initializeControlEnvironment({ config })); // ensure db is provisioned and get connection
    });
    afterAll(async () => {
      await connection.end();
    });
    it('should have an expected appearance when syncing change log for a change definition', async () => {
      // run the test
      stdout.stripColor = false; // dont strip color
      // stdout.print = true;
      stdout.start();
      await Sync.run([
        '-c',
        `${__dirname}/../__test_assets__/mysql/control.yml`,
        '--id',
        'init_service_user',
      ]);
      stdout.stop();
      const output = stdout.output
        .split('\n')
        .filter((line) => !line.includes('console.log'))
        .join('\n') // strip the console log portion
        .replace(/\[\d\d:\d\d:\d\d\]/g, ''); // remove all timestamps, since they change over time...
      expect(output).toMatchSnapshot();
    });
  });
  describe('postgres', () => {
    let connection: DatabaseConnection;
    beforeAll(async () => {
      const config = new ControlConfig({
        language: DatabaseLanguage.POSTGRES,
        dialect: '10.7',
        connection: await promiseConfigPostgres(),
        definitions: [],
        strict: false,
      });
      ({ connection } = await initializeControlEnvironment({ config })); // ensure db is provisioned and get connection
    });
    afterAll(async () => {
      await connection.end();
    });
    it('should have an expected appearance when syncing change log for a change definition', async () => {
      // run the test
      stdout.stripColor = false; // dont strip color
      // stdout.print = true;
      stdout.start();
      await Sync.run([
        '-c',
        `${__dirname}/../__test_assets__/postgres/control.yml`,
        '--id',
        'init_service_user',
      ]);
      stdout.stop();
      const output = stdout.output
        .split('\n')
        .filter((line) => !line.includes('console.log'))
        .join('\n') // strip the console log portion
        .replace(/\[\d\d:\d\d:\d\d\]/g, ''); // remove all timestamps, since they change over time...
      expect(output).toMatchSnapshot();
    });
  });
});
