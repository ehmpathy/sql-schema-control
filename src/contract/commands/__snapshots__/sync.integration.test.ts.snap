// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sync mysql should have an expected appearance when syncing change log for a change definition 1`] = `
"    
       [32mâœ”[39m [1m[32m[SYNC][39m src/contract/__test_assets__/mysql/definitions/init/service_user.sql [90m(change:init_service_user)[39m[22m
[2m[22m
[2m      [2mat [22m[2msrc/logic/commands/getAndSyncChangeLogForChangeDefinition.ts[2m:45:11[22m[2m[22m

"
`;

exports[`sync postgres should have an expected appearance when syncing change log for a change definition 1`] = `
"    
       [32mâœ”[39m [1m[32m[SYNC][39m src/contract/__test_assets__/postgres/definitions/init/service_user.sql [90m(change:init_service_user)[39m[22m
[2m[22m
[2m      [2mat [22m[2msrc/logic/commands/getAndSyncChangeLogForChangeDefinition.ts[2m:45:11[22m[2m[22m

"
`;
